ckanext/datagovuk/upload.py

```diff
@@ -0,0 +1,150 @@
+                            aws_access_key_id=aws_access_key_id,
+                            aws_secret_access_key=aws_secret_access_key)
+
+    bucket_name = config.get('ckan.datagovuk.s3_bucket_name')
+    bucket = s3.Bucket(bucket_name)
+
+    return bucket
+
+
+def upload_resource_to_s3(context, resource):
+    '''
+    upload_resource_to_s3
+
+    Uploads resource to S3 and modifies the following resource fields:
+    - 'upload'
+    - 'url_type'
+    - 'url'
+    '''
+
+    # Init logger
+    logger = logging.getLogger(__name__)
+    logger.info("Starting upload_resource_to_s3 for resource %s" % resource.get('name', ''))
+
+    # Init connection to S3
+    bucket = setup_s3_bucket()
+
+    # Get content type and extension
+    content_type, _ = mimetypes.guess_type(
+        resource.get('url', ''))
+    extension = mimetypes.guess_extension(content_type)
+
+    # Upload to S3
+    timestamp = datetime.datetime.utcnow()
+    pkg = toolkit.get_action('package_show')(context, {'id': resource['package_id']})
+    s3_filepath = (pkg.get('name')
+                   + '/'
+                   + 'resources'
+                   + '/'
+                   + resource.get('timestamp', timestamp.strftime("%Y-%m-%dT%H-%M-%SZ"))
+                   + '-'
+                   + slugify(resource.get('name'), to_lower=True)
+                   + extension)
+
+    # If file is currently being uploaded, the file is in resource['upload']
+    if isinstance(resource.get('upload', None), cgi.FieldStorage):
+        logger.info("File is being uploaded")
+        resource['upload'].file.seek(0)
+        body = resource['upload'].file
+    # If resource.get('url_type') == 'upload' then the resource is in CKAN file system
+    elif resource.get('url_type') == 'upload':
+        logger.info("File is on CKAN file store")
+        upload = uploader.ResourceUpload(resource)
+        filepath = upload.get_path(resource['id'])
+        try:
+            body = open(filepath, 'r')
+        except OSError:
+            abort(404, _('Resource data not found'))
+    else:
+        return ## in datagovuk, we don't want to upload from URL
+
+    try:
+        logger.info("Uploading resource %s to S3" % resource.get('name', ''))
+        bucket.Object(s3_filepath).delete()
+        obj = bucket.put_object(Key=s3_filepath,
+                                Body=body,
+                                ContentType=content_type)
+        obj.Acl().put(ACL='public-read')
+        logger.info("Successfully uploaded resource %s to S3" % resource.get('name', ''))
+
+    except ClientError as exception:
+        # Log the error and reraise the exception
+        logger.error("Error uploading resource %s from package %s to S3" % (resource['name'], resource['package_id']))
+        logger.error(exception)
+        if resource.get('url_type') == 'upload':
+            body.close()
+        raise exception
+
+    if resource.get('url_type') == 'upload':
+        body.close()
```

Comment:
You could [use a `finally` block](https://docs.python.org/3/tutorial/errors.html#defining-clean-up-actions) on the `try` here to avoid the duplication.

```python
finally:
    if resource.get('url_type') == 'upload':
        body.close()
```